presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayJoin.java

```diff
@@ -142,16 +141,16 @@ private static FunctionInfo specializeArrayJoin(Map<String, Type> types, Functio
         return new FunctionInfo(signature, DESCRIPTION, false, methodHandle.bindTo(castFunction).bindTo(type), true, false, nullableArguments);
     }
 
-    public static Slice arrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Slice array, Slice delimiter)
+    public static Slice arrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Block arrayBlock, Slice delimiter)
     {
-        return arrayJoin(castFunction, elementType, session, array, delimiter, null);
+        return arrayJoin(castFunction, elementType, session, arrayBlock, delimiter, null);
     }
 
-    public static Slice arrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Slice array, Slice delimiter, Slice nullReplacement)
+    public static Slice arrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Block arrayBlock, Slice delimiter, Slice nullReplacement)
     {
-        Block arrayBlock = readStructuralBlock(array);
         int numElements = arrayBlock.getPositionCount();
-        DynamicSliceOutput sliceOutput = new DynamicSliceOutput(array.length());
+
+        DynamicSliceOutput sliceOutput = new DynamicSliceOutput(arrayBlock.getPositionCount() * 5);
```

Comment:
`arrayBlock.getSizeInBytes() + delimiter.length() * arrayBlock.getPositionCount()` is probably a better estimate of the size